ingest.go

```diff
@@ -700,6 +705,109 @@ func (d *DB) IngestWithStats(paths []string) (IngestOperationStats, error) {
 	return d.ingest(paths, ingestTargetLevel)
 }
 
+// Both DB.mu and commitPipeline.mu must be held while this is called.
+func (d *DB) newIngestedSSTableFlushableEntry(
+	meta []*fileMetadata, seqNum uint64, logNum FileNum,
+) (*flushableEntry, error) {
+	// Update the sequence number for all of the sstables in the
+	// metadata. Writing the metadata to the manifest when the
+	// version edit is applied is the mechanism that persists the
+	// sequence number. The sstables themselves are left unmodified.
+	// In this case, a version edit will only be written to the manifest
+	// when the flushable is eventually flushed. If Pebble restarts in that
+	// time, then we'll lose the ingest sequence number information. But this
+	// information will also be reconstructed on node restart.
+	if err := ingestUpdateSeqNum(
+		d.cmp, d.opts.Comparer.FormatKey, seqNum, meta,
+	); err != nil {
+		return nil, err
+	}
+
+	f := newIngestedSSTableFlushable(meta, d.cmp, d.split, d.newIters, d.tableNewRangeIter)
+
+	// NB: The logNum/seqNum are the WAL number which we're writing this entry
+	// to and the sequence number within the WAL which we'll write this entry
+	// to.
+	entry := d.newFlushableEntry(f, logNum, seqNum)
+	// The flushable entry starts off with a single reader ref, so increment
+	// the FileMetadata.Refs.
+	for _, file := range f.files {
+		atomic.AddInt32(&file.Refs, 1)
+	}
+	entry.unrefFiles = func() []*fileMetadata {
+		var obsolete []*fileMetadata
+		for _, file := range f.files {
+			if val := atomic.AddInt32(&file.Refs, -1); val == 0 {
+				obsolete = append(obsolete, file)
+			}
+		}
+		return obsolete
+	}
+
+	entry.flushForced = true
+	entry.releaseMemAccounting = func() {}
+	return entry, nil
+}
+
+// Both DB.mu and commitPipeline.mu must be held while this is called.
+func (d *DB) handleIngestAsFlushable(paths []string, meta []*fileMetadata, seqNum uint64) error {
+	// If the WAL is disabled, then the logNum used to create the flushable
+	// entry doesn't matter. We just use the logNum assigned to the current
+	// flushable entry. If the WAL is enabled, then this logNum will be
+	// overwritten by the logNum of the log which will contain the log entry
+	// for the ingested sstables.
+	logNum := d.mu.mem.queue[len(d.mu.mem.queue)-1].logNum
+
+	b := d.NewBatch()
+	for _, path := range paths {
+		b.IngestSST([]byte(path))
+	}
+	b.setSeqNum(seqNum)
+
+	if !d.opts.DisableWAL {
+		// We create a new WAL for the flushable instead of reusing the end of
+		// the previous WAL. This simplifies the increment of the minimum
+		// unflushed log number, and also simplifies WAL replay.
+		logNum, _ = d.recycleWAL()
```

Comment:
This worries me -- we could have many ingests in quick succession and switching to a new WAL each time seems wasteful. Also,  I assume this is not needed for correctness, yes? What is the performance implication if we write the batch to the current WAL? Also the `logRecycler` can start keeping a bunch of small (even variable size is potentially bad) log files, which will cause more calls to `fdatasync` to also have to sync the file length metadata.

Reply:
Hmm, even if we don't start a new WAL, since we are creating a new memtable it will need to start a new WAL. But at least we save a tiny WAL that will only have this flushable entry in it.
I guess the variable sized WALs due to ingests causing memtable flushes is a potential problem even with our current solution. There should be some improvements possible in the `logRecycler` *if* this is a real problem.